#Theory Questions

### What is the difference between a function and a method in Python?

Function: A block of reusable code that operates independently and is not tied to any specific object. It is defined using def and can be called directly.
    
Method: A function that is associated with an object or class. It is called on an instance of a class and can access the instance's data.

### Explain the concept of function arguments and parameters in Python.


Parameters: These are the variables listed in a function's definition to accept values. They act as placeholders.  
  
Arguments: These are the actual values passed to a function when it is called.

### What are the different ways to define and call a function in Python?

In [ ]:
def greet(name):
    return f"Hello, {name}!"

print(greet("Alice"))

Hello, Alice!


In [ ]:
def greet(name="Guest"):
    return f"Hello, {name}!"

print(greet())        # Uses default value
print(greet("Alice")) # Uses provided value

Hello, Guest!
Hello, Alice!


In [ ]:
def add_numbers(*args):
    return sum(args)

print(add_numbers(1, 2, 3, 4))  # Any number of arguments

10


### What is the purpose of the return statement in a Python function?

If return is not used, the function returns None by default. It specifies the result of the function.

In [ ]:
def add(a, b):
    result = a + b
    return result

sum_result = add(5, 3)
print(sum_result)

8


### What are iterators in Python and how do they differ from iterables?

An object that can be looped over (e.g., lists, strings, tuples).  
It must implement the __iter__() method, which returns an iterator.  

An object that produces one value at a time using the __next__() method.  
It is created by calling iter() on an iterable.

In [ ]:
my_list = [1, 2, 3]  # List is an iterable
for item in my_list:
    print(item)

1
2
3


### Explain the concept of generators in Python and how they are defined.

In [ ]:
my_list = [1, 2, 3]
iterator = iter(my_list)  # Creating an iterator
print(next(iterator))  # Outputs: 1
print(next(iterator))  # Outputs: 2

1
2


### What are the advantages of using generators over regular functions?

Generators are memory-efficient as they produce items one at a time using `yield`, avoiding storing all values at once. They use lazy evaluation, computing values only when needed. Generators are simpler to implement than regular iterators and can handle infinite sequences easily. They are ideal for processing large datasets or streams.

In [ ]:
my_list = [1, 2, 3]
iterator = iter(my_list)
print(next(iterator))
print(next(iterator))

1
2


### What is a lambda function in Python and when is it typically used?

A lambda function in Python is a small, anonymous function defined using the lambda keyword. It can have multiple arguments but only one expression, which is implicitly returned.

In [ ]:
square = lambda x: x ** 2
print(square(5))

25


### Explain the purpose and usage of the map() function in Python.

The map() function in Python applies a given function to each item in an iterable (like a list or tuple) and returns a map object (an iterator).

In [ ]:
nums = [1, 2, 3, 4]
squares = map(lambda x: x**2, nums)
print(list(squares))

[1, 4, 9, 16]


### What is the difference between map(), reduce(), and filter() functions in Python?

map(): Transforms each element.
reduce(): Reduces the iterable to a single value.
filter(): Filters elements based on a condition.

In [ ]:
nums = [1, 2, 3]
result = map(lambda x: x ** 2, nums)
print(list(result))

[1, 4, 9]


### Using pen & paper, write the internal mechanism for the sum operation using the reduce() function on the given list: [47, 11, 42, 13].

In [ ]:
from functools import reduce

# List of numbers
numbers = [47, 11, 42, 13]

# Define the sum operation using lambda
result = reduce(lambda x, y: x + y, numbers)

# Output the result
print("The result of the sum operation is:", result)

The result of the sum operation is: 113


Internal Mechanism of the reduce() Function for Sum Operation
Given the list [47, 11, 42, 13], the reduce() function applies the sum operation step by step. Here's how it works:

Initial List:
[47, 11, 42, 13]

First Iteration:
The first two elements, 47 and 11, are passed to the function:

47 + 11 = 58
The result 58 is now used as the accumulator for the next iteration.

Second Iteration:
The accumulator 58 is combined with the next element 42:

58 + 42 = 100
The result 100 is now used for the next iteration.

Third Iteration:
The accumulator 100 is combined with the final element 13:

100 + 13 = 113
This is the final result.

Final Result:
After all iterations, the result is 113.

#Practical Questions

Here is the rewritten list of programming tasks:

1. Write a Python function that takes a list of numbers as input and returns the sum of all even numbers in the list.


In [ ]:
def sum_of_evens(numbers):
    return sum(x for x in numbers if x % 2 == 0)

sum_of_evens([1, 2, 3, 4, 5, 6])

12

2. Create a Python function that accepts a string and returns the reverse of that string.

In [ ]:
def reverse_string(s):
    return s[::-1]
reverse_string("hello")

'olleh'

3. Implement a Python function that takes a list of integers and returns a new list containing the squares of each number.


In [ ]:
def square_numbers(numbers):
    return [x ** 2 for x in numbers]

square_numbers([1, 2, 3, 4, 5])

[1, 4, 9, 16, 25]

4. Write a Python function that checks if a given number is prime or not from 1 to 200.

In [ ]:
def is_prime(n):
    if n < 2:
        return False
    for i in range(2, int(n ** 0.5) + 1):
        if n % i == 0:
            return False
    return True

primes = [n for n in range(1, 201) if is_prime(n)]

is_prime(17)

True

5. Create an iterator class in Python that generates the Fibonacci sequence up to a specified number of terms.

In [ ]:
class FibonacciIterator:
    def __init__(self, terms):
        self.terms = terms
        self.a, self.b = 0, 1
        self.count = 0

    def __iter__(self):
        return self

    def __next__(self):
        if self.count < self.terms:
            result = self.a
            self.a, self.b = self.b, self.a + self.b
            self.count += 1
            return result
        else:
            raise StopIteration


fib = FibonacciIterator(10)
for num in fib:
    print(num)

0
1
1
2
3
5
8
13
21
34


6. Write a generator function in Python that yields the powers of 2 up to a given exponent.

In [ ]:
def powers_of_two(exponent):
    for i in range(exponent + 1):
        yield 2 ** i

# Example usage:
for power in powers_of_two(5):
    print(power)

1
2
4
8
16
32


7. Implement a generator function that reads a file line by line and yields each line as a string.

In [ ]:
def read_file_line_by_line(file_path):
    with open(file_path, 'r') as file:
        for line in file:
            yield line.strip()

# Example usage:
# for line in read_file_line_by_line('example.txt'):
#     print(line)

8. Use a lambda function in Python to sort a list of tuples based on the second element of each tuple.

In [ ]:
tuples = [(1, 3), (2, 2), (3, 1), (4, 4)]

sorted_tuples = sorted(tuples, key=lambda x: x[1])

print(sorted_tuples)

[(3, 1), (2, 2), (1, 3), (4, 4)]


9. Write a Python program that uses `map()` to convert a list of temperatures from Celsius to Fahrenheit.

In [ ]:
celsius_temperatures = [0, 20, 37, 100]

fahrenheit_temperatures = list(map(lambda c: (c * 9/5) + 32, celsius_temperatures))

print(fahrenheit_temperatures)

[32.0, 68.0, 98.6, 212.0]


10. Create a Python program that uses `filter()` to remove all the vowels from a given string.

In [ ]:
def remove_vowels(input_string):
    vowels = "aeiouAEIOU"
    return ''.join(filter(lambda x: x not in vowels, input_string))

input_string = "Hello, World!"
result = remove_vowels(input_string)

print(result)

Hll, Wrld!
